Start amount organization specification (#88)
zkSNACKs/WabiSabi: amounts.md

# Amount Organization Specification for Wasabi Wallet 2.0

## Target User

Peter McCormack

## Objectives

- Avoid homogenic coin distribution.
- Avoid UTXO bloat.
- Significantly lower minimum denominations compared to [WW1.](https://github.com/nopara73/ZeroLink/)
- Thrive for not generating non-mixed changes.
- Thrive for faster mixing.
- Thrive for more blockspace efficient mixing.

## Approach

Since this amount organization will be used in Wasabi Wallet 2.0, a sensible strategy is to start out with the current amount organization of WW1 and step by step improve upon that.

### Setup

In order to verify our reasoning and intuitions we [sampled 8000 input amounts](./AmountOrganization/AmountOrganization/Sample.txt) from Wasabi Wallet 1.0 coinjoins around December 2020, those have not been mixed before.  

> We'll use in various ways throughout this document, mostly from code, which source can be found under [AmountOrganization solution.](./AmountOrganization)

Reviewer: This sentence is not clear to me, what will be used?
Do you want to say `We'll use them ...`?

#### BASE: WWI Skeleton

> To start out, we took 50 inputs randomly out of our sample and created 40 random groupings - users - out of it. Then implemented WW1's mixing technique and got the following example output.

Reviewer: ```suggestion
To start out, from our sample we took a set of 50 inputs randomly and created 40 random groupings - users - out of it. Then we implemented WW1's mixing technique and got the following example output.
```

There are 23 occurrences of 0.02097152 BTC output.
There are 1 occurrences of 0.03033266 BTC output.
There are 1 occurrences of 0.0393451 BTC output.
There are 14 occurrences of 0.04194304 BTC output.
There are 1 occurrences of 0.05653016 BTC output.
There are 1 occurrences of 0.07940192 BTC output.
There are 7 occurrences of 0.08388608 BTC output.
There are 1 occurrences of 0.11869243 BTC output.
There are 1 occurrences of 0.14460622 BTC output.
There are 5 occurrences of 0.16777216 BTC output.
There are 1 occurrences of 0.2086305 BTC output.
There are 4 occurrences of 0.33554432 BTC output.
There are 1 occurrences of 0.65823 BTC output.
There are 3 occurrences of 0.67108864 BTC output.
There are 2 occurrences of 1.34217728 BTC output.
There are 2 occurrences of 2.68435456 BTC output.
There are 1 occurrences of 4.72813231 BTC output.
There are 1 occurrences of 5.36870912 BTC output.
There are 1 occurrences of 10.73741824 BTC output.
```

#### Descending Powers of 2

It would be more efficient to do descending powers of 2 mixing instead of ascending powers of 2 mixing like we did before. So we implemented this and here's what we noticed:

> - Now more dust coins would be created, but because of their low economic value, these coins are not created and instead the value is added to the miner fees. Note that dust warnings are ommitted on the output pasted here.

Reviewer: ```suggestion
- Now more dust coins would be created, but because of their low economic value, these coins are not created and instead their value is added to the miner fees. Note that dust warnings are omitted on the output pasted here.
```

> - We're creating roughly half the number of outputs, which is gret for blockspace.  

Reviewer: ```suggestion
- We're creating roughly half the number of outputs, which is great for blockspace.  
```

- We've nearly completely eliminated change.  

Furthermore a few more small optimization was introduced:

- Our sanity dust threshold became 1024, so we won't create smaller change than that.  

> - A user now also consider the largest output that isn't her so she won't create denomination that would be larger than the largest possible user's output she could reasonably mix with.

Reviewer: ```suggestion
- A user now also considers the largest output that isn't hers so she won't create denomination that would be larger than the largest possible user's output she could reasonably mix with.
```

Size:                   12657 vbyte
Fee rate:               11 sats/vbyte

There are 17 occurrences of     0.00001024 BTC output.
There are 15 occurrences of     0.00002048 BTC output.
There are 22 occurrences of     0.00004096 BTC output.
There are 24 occurrences of     0.00008192 BTC output.
There are 22 occurrences of     0.00016384 BTC output.
There are 23 occurrences of     0.00032768 BTC output.
There are 22 occurrences of     0.00065536 BTC output.
There are 20 occurrences of     0.00131072 BTC output.
There are 21 occurrences of     0.00262144 BTC output.
There are 20 occurrences of     0.00524288 BTC output.
There are 16 occurrences of     0.01048576 BTC output.
There are 12 occurrences of     0.02097152 BTC output.
There are 15 occurrences of     0.04194304 BTC output.
There are 11 occurrences of     0.08388608 BTC output.
There are 6 occurrences of      0.16777216 BTC output.
There are 6 occurrences of      0.33554432 BTC output.
There are 3 occurrences of      0.67108864 BTC output.
There are 3 occurrences of      1.34217728 BTC output.
There are 1 occurrences of      1.91315182 BTC output.
```

#### Should we break down the largest user's bag?

> A user whose combined input amount is significantly larger (or smaller) than the other users always face difficulties in mixing. 

Reviewer: ```suggestion
A user whose combined input amount is significantly larger (or smaller) than the other users' amounts always faces difficulties in mixing.
```

> In this subsection we address different approaches to managing large outputs amounts. Smaller amounts are not addressed here.

Reviewer: ```suggestion
In this subsection we address different approaches to managing large output amounts. Smaller amounts are not addressed here.
```

We identified five possible approaches. 

The trivial approach is not to break the large input(s) up. 
We just produce a large amount change minus the fees. This output is deterministically linkable to its input cluster.
Potentially, it might also reveal what other amounts were used by the user.

Next, we consider different ways of splitting up the large amounts. 
If these amounts are split up too granularly, then a single user with more liquidity than others is indistinguishable from a sybil attack.
Additionally, this user gains little privacy themselves (probabilistic privacy instead of k-anonymity). 
Moreover, this approach is not blockspace efficient.
On the other hand, if they are split up too coarsely, then that user's amounts are deterministically linkable to their inputs.

We discuss the following approaches for splitting up the large amount inputs:

> The following design decisions needs to be made. 

Reviewer: ```suggestion
The following design decisions need to be made.
```

# Amount Organization Specification for Wasabi Wallet 2.0

## Target User

Peter McCormack

## Objectives

- Avoid homogenic coin distribution.
- Avoid UTXO bloat.
- Significantly lower minimum denominations compared to [WW1.](https://github.com/nopara73/ZeroLink/)
- Thrive for not generating non-mixed changes.
- Thrive for faster mixing.
- Thrive for more blockspace efficient mixing.

## Approach

Since this amount organization will be used in Wasabi Wallet 2.0, a sensible strategy is to start out with the current amount organization of WW1 and step by step improve upon that.

### Setup

In order to verify our reasoning and intuitions we [sampled 8000 input amounts](./AmountOrganization/AmountOrganization/Sample.txt) from Wasabi Wallet 1.0 coinjoins around December 2020, those have not been mixed before.  
We'll use in various ways throughout this document, mostly from code, which source can be found under [AmountOrganization solution.](./AmountOrganization)

#### BASE: WWI Skeleton

To start out, we took 50 inputs randomly out of our sample and created 40 random groupings - users - out of it. Then implemented WW1's mixing technique and got the following example output.

> How does WW1's mixing technique work? It takes the amounts and loops them through powers of 2 denominations in ascending order. It creates outputs for every denomination level the user can still participate in. If it can't it creates a change output.

Reviewer 2: ```suggestion
How does WW1's mixing technique work? It takes the amounts and loops them through powers of 2 denominations in ascending order. It creates outputs for every denomination level the user can still participate in. If it can't, it creates a change output.
```

There are 1 occurrences of 0.05586343 BTC output.
There are 1 occurrences of 0.06800123 BTC output.
There are 1 occurrences of 0.06908003 BTC output.
There are 12 occurrences of 0.08388608 BTC output.
There are 1 occurrences of 0.10004366 BTC output.
There are 1 occurrences of 0.10557624 BTC output.
There are 1 occurrences of 0.10746535 BTC output.
There are 1 occurrences of 0.11578316 BTC output.
There are 1 occurrences of 0.13324562 BTC output.
There are 1 occurrences of 0.1543588 BTC output.
There are 7 occurrences of 0.16777216 BTC output.
There are 4 occurrences of 0.33554432 BTC output.
There are 2 occurrences of 0.67108864 BTC output.
```

#### Dust Strategy

Our next thought was to make a drastic change on the amount organization structure, but we realized the nature of the change may depend on how fees are handled and that may depend on our dust handling strategy, thus it's reasonable to continue with that.  
So far we've created a transaction that isn't standard, because many outputs, like one satoshi outputs are present. To manage this issue and make our transaction standard, we want to make sure users don't register small outputs by setting a dust limit.  

> Dismissing the dust by the user is fine, because it's not significant amount. Let's assume our dust limit is 1000 satoshis, which should satisfy the standardness rules.

Reviewer 2: ```suggestion
Dismissing the dust by the user is fine, because it's not significant amount. Let's assume our dust limit is 1023 satoshis, which should satisfy the standardness rules.
```

we don't want to have dust below the last standard amount, no?

not sure if my change should be `1023` or `1024`.

Reviewer 2: I see this is discussed later in `Descending Powers of 2`.

But then it might be worth to mention here that we had small coins below a standard denomination, which was a problem. then later we tell how we fixed the problem.

Author: At this step the dust is 1000, we changed it later.

There are 1 occurrences of 0.03322046 BTC output.
There are 1 occurrences of 0.03623939 BTC output.
There are 1 occurrences of 0.03748776 BTC output.
There are 15 occurrences of 0.04194304 BTC output.
There are 1 occurrences of 0.04467416 BTC output.
There are 9 occurrences of 0.08388608 BTC output.
There are 1 occurrences of 0.10879512 BTC output.
There are 1 occurrences of 0.11102965 BTC output.
There are 9 occurrences of 0.16777216 BTC output.
There are 1 occurrences of 0.2740125 BTC output.
There are 1 occurrences of 0.31065568 BTC output.
There are 2 occurrences of 0.3289216 BTC output.
There are 8 occurrences of 0.33554432 BTC output.
There are 1 occurrences of 0.54129616 BTC output.
There are 3 occurrences of 0.67108864 BTC output.
There are 1 occurrences of 0.757806 BTC output.
There are 1 occurrences of 0.98800148 BTC output.
There are 1 occurrences of 1.34217728 BTC output.
```

Although we handled our standardness problem, we also want to make sure we don't create uneconomical outputs. Our working formula for this is: `MAX((feerate * input spend size estimation), 1000, (minimum relay fee rate * input spend size estimation))`.

> If we choose the current fee rate to 10s/vb, and estimate our input spend size to 69 bytes, then we'll still be under the 1000 satoshi sanity fee rate, so we don't need to illustrate our new results.

Reviewer 2: ```suggestion
If we choose the current fee rate to 10s/vb, and estimate our input spend size to 69 bytes, then we'll still be under the 1023 satoshi sanity fee rate, so we don't need to illustrate our new results.
```

if [the above](https://github.com/zkSNACKs/WabiSabi/pull/88/files#r540028543) gets merged.

Author: Same comment. The dust change is mentioned down the line.

There are 23 occurrences of 0.02097152 BTC output.
There are 1 occurrences of 0.03033266 BTC output.
There are 1 occurrences of 0.0393451 BTC output.
There are 14 occurrences of 0.04194304 BTC output.
There are 1 occurrences of 0.05653016 BTC output.
There are 1 occurrences of 0.07940192 BTC output.
There are 7 occurrences of 0.08388608 BTC output.
There are 1 occurrences of 0.11869243 BTC output.
There are 1 occurrences of 0.14460622 BTC output.
There are 5 occurrences of 0.16777216 BTC output.
There are 1 occurrences of 0.2086305 BTC output.
There are 4 occurrences of 0.33554432 BTC output.
There are 1 occurrences of 0.65823 BTC output.
There are 3 occurrences of 0.67108864 BTC output.
There are 2 occurrences of 1.34217728 BTC output.
There are 2 occurrences of 2.68435456 BTC output.
There are 1 occurrences of 4.72813231 BTC output.
There are 1 occurrences of 5.36870912 BTC output.
There are 1 occurrences of 10.73741824 BTC output.
```

#### Descending Powers of 2

It would be more efficient to do descending powers of 2 mixing instead of ascending powers of 2 mixing like we did before. So we implemented this and here's what we noticed:

> - Now there are many dusts would had been created, but they don't matter, because they're cents and can just be swallowed by the miners.  

Reviewer 2: ```suggestion
- Now more dust coins would be created, but because of their low economic value, these coins are not created and instead the value is added to the miner fees.  
```

> - We're creating roughly half the number of outputs, which is gret for blockspace.  

Reviewer 2: ```suggestion
- We're creating roughly half the number of outputs, which is great for blockspace.  
```
